solana/program_v2/src/NetworkService.ts

```diff
@@ -0,0 +1,270 @@
+import * as anchor from '@project-serum/anchor';
+import { AnchorProvider, Program } from '@project-serum/anchor';
+import { ConfirmOptions, PublicKey } from '@solana/web3.js';
+import {
+  AuthKeyStructure,
+  CreateGatekeeperData,
+  UpdateGatekeeperData,
+  FeeStructure,
+  Wallet,
+  GatekeeperAccount,
+  GatekeeperState,
+  GatekeeperStateMapping,
+} from './lib/types';
+import {
+  CustomClusterUrlConfig,
+  ExtendedCluster,
+  getConnectionByCluster,
+} from './lib/connection';
+import { EnumMapper } from './lib/utils';
+import {
+  GatekeeperKeyFlags,
+  GATEKEEPER_SEED,
+  GATEWAY_PROGRAM,
+  SOLANA_MAINNET,
+} from './lib/constants';
+import { AbstractService, ServiceBuilder } from './utils/AbstractService';
+import { GatewayV2 } from '../target/types/gateway_v2';
+
+// Service for a network. This will handle all aspects of the Gateway that a network is able to control... i.e. creating gatekeepers, updating gatekeepers, etc...
+export class NetworkService extends AbstractService {
+  static async build(
+    dataAccount: PublicKey,
+    wallet: Wallet,
+    cluster: ExtendedCluster = SOLANA_MAINNET,
+    customConfig?: CustomClusterUrlConfig,
+    opts: ConfirmOptions = AnchorProvider.defaultOptions()
+  ): Promise<NetworkService> {
+    const _connection = getConnectionByCluster(
+      cluster,
+      opts.preflightCommitment,
+      customConfig
+    );
+
+    const provider = new AnchorProvider(_connection, wallet, opts);
+
+    const program = await NetworkService.fetchProgram(provider);
+
+    return new NetworkService(
+      program,
+      dataAccount,
+      cluster,
+      wallet,
+      provider.opts
+    );
+  }
+
+  static async buildFromAnchor(
+    program: Program<GatewayV2>,
+    dataAccount: PublicKey,
+    cluster: ExtendedCluster,
+    provider: AnchorProvider = program.provider as AnchorProvider,
+    wallet: Wallet = provider.wallet
+  ): Promise<NetworkService> {
+    return new NetworkService(
+      program,
+      dataAccount,
+      cluster,
+      wallet,
+      provider.opts
+    );
+  }
+
+  // Creates a gatekeeper's public key from a given seed and authority.
+  static async createGatekeeperAddress(
+    authority: PublicKey,
+    network: PublicKey
+  ): Promise<[PublicKey, number]> {
+    return PublicKey.findProgramAddress(
+      [
+        anchor.utils.bytes.utf8.encode(GATEKEEPER_SEED),
+        authority.toBuffer(),
+        network.toBuffer(),
+      ],
+      GATEWAY_PROGRAM
+    );
+  }
+
+  // Creates a gatekeeper on a specified network
+  // @authThreshold
+  createGatekeeper(
+    network: PublicKey,
+    data: CreateGatekeeperData = {
+      gatekeeperBump: 0,
+      gatekeeperNetwork: network,
+      // TODO? Is this the correct way to derive the default staking account?
+      stakingAccount: PublicKey.findProgramAddressSync(
```

Comment:
Can we pull this out into a Staking Account interface? This should not be done here.